java/client/src/main/java/glide/connectors/handlers/MessageHandler.java

```diff
@@ -152,4 +154,62 @@ enum PushKind {
         /// `ssubscribe` is received when client subscribed to a shard channel.
         SSubscribe,
     }
+
+    /**
+     * An asynchronous FIFO message queue for {@link PubSubMessage} backed by {@link
+     * ConcurrentLinkedDeque}.
+     */
+    public static class PubSubMessageQueue {
+        /** The queue itself. */
+        final ConcurrentLinkedDeque<PubSubMessage> messageQueue = new ConcurrentLinkedDeque<>();
+
+        /**
+         * A promise for the first incoming message. Returned to a user, if message queried in async
+         * manner, but nothing received yet.
+         */
+        CompletableFuture<PubSubMessage> firstMessagePromise = new CompletableFuture<>();
+
+        /** A flag whether a user already got a {@link #firstMessagePromise}. */
+        private final AtomicBoolean firstMessagePromiseRequested = new AtomicBoolean(false);
```

Comment:
Just needs to be a regular boolean since this is always used within synchronized blocks.